cli/internal/cmd/miniup.go

```diff
@@ -0,0 +1,268 @@
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+package cmd
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"runtime"
+	"strings"
+
+	"github.com/edgelesssys/constellation/v2/cli/internal/cloudcmd"
+	"github.com/edgelesssys/constellation/v2/cli/internal/helm"
+	"github.com/edgelesssys/constellation/v2/cli/internal/libvirt"
+	"github.com/edgelesssys/constellation/v2/internal/cloud/cloudprovider"
+	"github.com/edgelesssys/constellation/v2/internal/config"
+	"github.com/edgelesssys/constellation/v2/internal/constants"
+	"github.com/edgelesssys/constellation/v2/internal/file"
+	"github.com/edgelesssys/constellation/v2/internal/grpc/dialer"
+	"github.com/edgelesssys/constellation/v2/internal/license"
+	"github.com/edgelesssys/constellation/v2/internal/versions"
+	"github.com/schollz/progressbar/v3"
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+)
+
+func newMiniUpCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "up",
+		Short: "Create and initialize a new mini Constellation cluster",
+		Long: "Create and initialize a new mini Constellation cluster.\n" +
+			"A mini cluster consists of a single control-plane and worker node, hosted using QEMU/KVM.\n",
+		Args: cobra.ExactArgs(0),
+		RunE: runUp,
+	}
+
+	// override global flag so we don't have a default value for the config
+	cmd.Flags().String("config", "", "path to the config file to use for the cluster")
+
+	return cmd
+}
+
+func runUp(cmd *cobra.Command, args []string) error {
+	if err := checkSystemRequirements(cmd.OutOrStdout()); err != nil {
+		return fmt.Errorf("system requirements not met: %w", err)
+	}
+
+	fileHandler := file.NewHandler(afero.NewOsFs())
+
+	// create config if not passed as flag and set default values
+	config, err := prepareConfig(cmd, fileHandler)
+	if err != nil {
+		return fmt.Errorf("preparing config: %w", err)
+	}
+
+	// create cluster
+	spinner := newSpinner(cmd, "Creating cluster in QEMU ", true)
+	spinner.Start()
+	err = createMiniCluster(cmd.Context(), fileHandler, cloudcmd.NewCreator(cmd.OutOrStdout()), config)
+	spinner.Stop()
+	if err != nil {
+		return fmt.Errorf("creating cluster: %w", err)
+	}
+	cmd.Println("Cluster successfully created.")
+	connectURI := config.Provider.QEMU.LibvirtURI
+	if connectURI == "" {
+		connectURI = libvirt.LibvirtTCPConnectURI
+	}
+	cmd.Println("Connect to the VMs by executing:")
+	cmd.Printf("\tvirsh -c %s\n\n", connectURI)
+
+	// initialize cluster
+	if err := initializeMiniCluster(cmd, fileHandler); err != nil {
+		return fmt.Errorf("initializing cluster: %w", err)
+	}
+	return nil
+}
+
+// checkSystemRequirements checks if the system meets the requirements for running a mini Constellation cluster.
+// We do so by verifying that the host:
+// - arch/os is linux/amd64.
+// - has access to /dev/kvm.
+// - has more than 4 CPU cores.
+// - has more than 4GB of memory.
```

Comment:
Already discussed, just as a reminder: Also check for free disk space available.